api/admin/controller/metadata_services.py

```diff
@@ -0,0 +1,137 @@
+from nose.tools import set_trace
+import flask
+from flask import Response
+from api.admin.problem_details import *
+from api.nyt import NYTBestSellerAPI
+from api.novelist import NoveListAPI
+from core.opds_import import MetadataWranglerOPDSLookup
+from core.model import (
+    ExternalIntegration,
+    get_one,
+)
+from core.util.http import HTTP
+from core.util.problem_detail import ProblemDetail
+
+from . import SettingsController
+
+class MetadataServicesController(SettingsController):
+
+    def __init__(self, manager):
+        super(MetadataServicesController, self).__init__(manager)
+        provider_apis = [
+                            NYTBestSellerAPI,
+                            NoveListAPI,
+                            MetadataWranglerOPDSLookup,
+                        ]
+
+        self.protocols = self._get_integration_protocols(provider_apis, protocol_name_attr="PROTOCOL")
+
+    def process_metadata_services(self):
+        self.require_system_admin()
+        if flask.request.method == 'GET':
+            return self.process_get()
+        else:
+            return self.process_post()
+
+    def process_get(self):
+        metadata_services = self._get_integration_info(ExternalIntegration.METADATA_GOAL, self.protocols)
+        return dict(
+            metadata_services=metadata_services,
+            protocols=self.protocols,
+        )
+
+    def process_post(self, do_get=HTTP.debuggable_get, do_post=HTTP.debuggable_post):
+        name = flask.request.form.get("name")
+        protocol = flask.request.form.get("protocol")
+        fields = {"name": name, "protocol": protocol}
+        form_field_error = self.validate_form_fields(**fields)
+        if form_field_error:
+            return form_field_error
+
+        id = flask.request.form.get("id")
+        is_new = False
+        if id:
+            # Find an existing service in order to edit it
+            service = self.look_up_service_by_id(id, protocol)
+        else:
+            service, is_new = self._create_integration(
+                self.protocols, protocol, ExternalIntegration.METADATA_GOAL
+            )
+
+        if isinstance(service, ProblemDetail):
+            self._db.rollback()
+            return service
+
+        error = self.check_name_unique(service, name) or self.set_protocols(service, protocol) or self.register_with_metadata_wrangler(is_new, service)
```

Comment:
check_name_unique is just a validation check, but set_protocols and register_with_metadata_wrangler have serious side effects. It's fine to use `or` to string together validation checks, but a method call that has side effects should be on its own line, even though this makes the code longer with an `if error:` check for each method call. Otherwise it's easy to overlook that the things with side effects are even happening.